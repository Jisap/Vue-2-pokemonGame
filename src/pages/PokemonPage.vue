<template>
<div>
  
  <h1 v-if="!pokemon">Espere por favor...</h1>

  <div v-else>
    <h1>¿ Quien es este pokemon ?</h1>
    <PokemonPicture 
      :pokemonId="pokemon.id"
      :showPokemon="showPokemon"/> <!--props del modelo PokemonPicture vinculadas aquí con los valores de la data-->
    <PokemonOptions
       :pokemons="pokemonArr"     
       @selection="checkAnswer"/> <!--Prop pokemons vinculada a la data de PokemonPage que recibe el [{}] -->
                                  <!--Aquí estaremos escuchando el evento selection que disparará el método checkAnswer-->
  
    <div v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">
        Nuevo Juego
      </button>
    </div>  
  
  </div>

</div>
</template>

<script>
import PokemonOptions from "../components/pokemonOptions.vue"
import PokemonPicture from "../components/PokemonPicture.vue"
import getPokemonOptions from '../helpers/getPokemonOptions'

//console.log( getPokemonOptions())

export default {
  components:{ PokemonOptions, PokemonPicture },
  data(){
    return {
      pokemonArr:[],
      pokemon:null,    // pokemon que la persona tiene que adivinar
      showPokemon: false,
      showAnswer: false,
      message: '',
    }
  },
  methods:{ 
    async mixPokemonArray(){
      
      this.pokemonArr = await getPokemonOptions(); // Asignamos el [{}] a la data de esta pokemonPage
      
      const rndInt = Math.floor( Math.random() * 4 ); // Nº aleatorio . Redondeamos un nº aleatorio entre 0,1,2 y 3
      this.pokemon = this.pokemonArr[ rndInt ]        // El pokemon a adivinar estará determinado por este nº aleatorio
    },
    checkAnswer( selectedId ){
      // Este método recibirá el contenido del evento (pokemon seleccionado por el jugador)
      this.showPokemon = true;
      this.showAnswer = true;

      if( selectedId === this.pokemon.id ){
        this.message = `Correcto, ${ this.pokemon.name }`
      }else{
        this.message = `Oops, era ${ this.pokemon.name }`
      }
    },
    newGame(){
      this.showPokemon = false,
      this.showAnswer = false,
      this.pokemonArr = [],
      this.pokemon = null,
      this.mixPokemonArray()
    }, 
  },
  mounted(){
    this.mixPokemonArray();// Cuando el componente se ha montado se llama a este método y se hace la asignacion del [{}] con la prop de pokemonOPtions
  }
}
</script>

