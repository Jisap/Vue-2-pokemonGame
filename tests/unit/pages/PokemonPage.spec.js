import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock';


describe('Pokemonpage Component', () => {

    let wrapper
    
    beforeEach(() => {
        wrapper = shallowMount( PokemonPage )
    })
    
    test('debe hacer match con el snapshot', () => {
        expect( wrapper.html()).toMatchSnapshot()
    });
    
    
    test('debe de llamar mixPokemonArray al montar', () => {
        console.log(PokemonPage.methods)
        const mixPokemonArraySpy = jest.spyOn( PokemonPage.methods , 'mixPokemonArray' )
        shallowMount( PokemonPage )
        expect( mixPokemonArraySpy ).toHaveBeenCalled()
    });
    
    test('debe de hacer match con el snapshot cuando cargan los pokemons', () => {
       const wrapper = shallowMount( PokemonPage, {
           data(){
               return {
                   pokemonArr: pokemons,
                   pokemon: pokemons[0],
                   showPokemon:false,
                   showAnswer:false,
                   message:'',
               }
           }
       });
       expect( wrapper.html()).toMatchSnapshot();
    });

    test('debe de mostrar los componentes de PokemonPicture y PokemonOptions', () => {
        
        const wrapper = shallowMount( PokemonPage, {
           data(){
               return {
                   pokemonArr: pokemons,
                   pokemon: pokemons[0],
                   showPokemon:false,
                   showAnswer:false,
                   message:'',
               }
           }
       });

       expect( wrapper.find('pokemon-picture-stub').exists()).toBeTruthy()
       expect( wrapper.find('pokemon-options-stub').exists()).toBeTruthy()
       expect( wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe("1")
       expect( wrapper.find('pokemon-options-stub').attributes('pokemons')).toBeTruthy()
       
    });
    
    test('pruebas con checkAnswer', async() => {
        
        const wrapper = shallowMount( PokemonPage, {    // Montamos el componente
           data(){
               return {
                   pokemonArr: pokemons,
                   pokemon: pokemons[0],//Bulbasaur id=1
                   showPokemon:false,
                   showAnswer:false,
                   message:'',
               }
           }
       });

       await wrapper.vm.checkAnswer(1);                     // Simulamos que la respuesta en las opciones es 1 (selectId)
       expect( wrapper.find('h2').exists()).toBeTruthy()    // Esperariamos que se encontrará en el wrapper un h2 respondiendo a la opción seleccionada
       console.log(wrapper.find('h2').text())               // Bulbasaur
       expect( wrapper.vm.showPokemon ).toBeTruthy()        // Esperariamos que la prop showPokemon = true () puesto que selectedId=pokemon[0]
       expect( wrapper.find('h2').text()).toBe(`Correcto, ${ pokemons[0].name}`) // Ademas el mensaje debería = 'Correcto, Bulbasaur'
    
       await wrapper.vm.checkAnswer(10)
       expect( wrapper.vm.message ).toBe(`Oops, era ${ pokemons[0].name }`)
    });
    
});
