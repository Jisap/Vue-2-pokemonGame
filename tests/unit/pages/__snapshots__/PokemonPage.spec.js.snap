// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemonpage Component debe de hacer match con el snapshot cuando cargan los pokemons 1`] = `
<div>
  <div>
    <h1>¿ Quien es este pokemon ?</h1>
    <pokemon-picture-stub pokemonid="1" showpokemon="false"></pokemon-picture-stub>
    <!--props del modelo PokemonPicture vinculadas aquí con los valores de la data-->
    <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
    <!--Prop pokemons vinculada a la data de PokemonPage que recibe el [{}] -->
    <!--Aquí estaremos escuchando el evento selection que disparará el método checkAnswer-->
    <!--v-if-->
  </div>
</div>
`;

exports[`Pokemonpage Component debe hacer match con el snapshot 1`] = `
<div>
  <h1>Espere por favor...</h1>
</div>
`;
